internal/resolution/entitysources/catalogdsource.go

```diff
@@ -71,70 +72,117 @@ func (es *CatalogdEntitySource) Iterate(ctx context.Context, fn input.IteratorFu
-						Replaces: b.Replaces,
-					})
-					props[entities.PropertyBundleChannelEntry] = string(replacesValue)
-					entity := input.Entity{
-						ID:         deppy.IdentifierFromString(fmt.Sprintf("%s%s%s", bundle.Name, bundle.Spec.Package, ch.Name)),
-						Properties: props,
-					}
-					entityList = append(entityList, entity)
+
+		allEntitiesList = append(allEntitiesList, catalogEntitiesList...)
+	}
+
+	return allEntitiesList, nil
+}
+
+func MetadataToEntities(catalogName string, channels []declcfg.Channel, bundles []declcfg.Bundle) (input.EntityList, error) {
+	entityList := input.EntityList{}
+
+	bundlesMap := map[string]*declcfg.Bundle{}
+	for i := range bundles {
+		bundlesMap[bundles[i].Name] = &bundles[i]
+	}
+
+	for _, ch := range channels {
+		for _, chEntry := range ch.Entries {
+			bundle, ok := bundlesMap[chEntry.Name]
+			if !ok {
+				return nil, fmt.Errorf("bundle %q not found in catalog %q (package %q, channel %q)", chEntry.Name, catalogName, ch.Package, ch.Name)
+			}
+
+			props := map[string]string{}
+
+			for _, prop := range bundle.Properties {
+				switch prop.Type {
+				case property.TypePackage:
+					// this is already a json marshalled object, so it doesn't need to be marshalled
+					// like the other ones
+					props[property.TypePackage] = string(prop.Value)
+				case entities.PropertyBundleMediaType:
+					props[entities.PropertyBundleMediaType] = string(prop.Value)
 				}
 			}
+
+			imgValue, err := json.Marshal(bundle.Image)
+			if err != nil {
+				return nil, err
+			}
+			props[entities.PropertyBundlePath] = string(imgValue)
+
+			channelValue, _ := json.Marshal(property.Channel{ChannelName: ch.Name, Priority: 0})
+			props[property.TypeChannel] = string(channelValue)
+			replacesValue, _ := json.Marshal(entities.ChannelEntry{
+				Name:     bundle.Name,
+				Replaces: chEntry.Replaces,
+			})
+			props[entities.PropertyBundleChannelEntry] = string(replacesValue)
+
+			catalogScopedEntryName := fmt.Sprintf("%s-%s", catalogName, bundle.Name)
+			entity := input.Entity{
+				ID:         deppy.IdentifierFromString(fmt.Sprintf("%s%s%s", catalogScopedEntryName, bundle.Package, ch.Name)),
+				Properties: props,
+			}
+			entityList = append(entityList, entity)
 		}
 	}
+
 	return entityList, nil
 }
 
-func fetchMetadata(ctx context.Context, client client.Client) (catalogd.BundleMetadataList, map[string]catalogd.Package, error) {
-	packageMetdatas := catalogd.PackageList{}
-	if err := client.List(ctx, &packageMetdatas); err != nil {
-		return catalogd.BundleMetadataList{}, nil, err
+func fetchCatalogMetadata(ctx context.Context, cl client.Client, catalogName string) ([]declcfg.Channel, []declcfg.Bundle, error) {
+	channels, err := fetchCatalogMetadataByScheme[declcfg.Channel](ctx, cl, declcfg.SchemaChannel, catalogName)
+	if err != nil {
+		return nil, nil, err
+	}
+	bundles, err := fetchCatalogMetadataByScheme[declcfg.Bundle](ctx, cl, declcfg.SchemaBundle, catalogName)
+	if err != nil {
+		return nil, nil, err
```

Comment:
Do we only care about channels and bundles? What about packages?

Reply:
We only need package name to create an entity and both channels and bundles have it. So for now channels and bundles are enough. But we might need to fetch packages at some point.